src/keys/public.rs

```diff
@@ -106,9 +99,123 @@ impl PublicKey {
     /// 2. It is on the curve.
     /// 3. It is not the identity.
     pub fn is_valid(&self) -> bool {
-        let is_identity: bool = self.0.is_identity().into();
-        self.0.is_torsion_free().into()
-            && self.0.is_on_curve().into()
-            && !is_identity
+        is_valid(&self.0)
+    }
+}
+
+fn verify(key: &G2Affine, sig: &G1Affine, msg: &[u8]) -> Result<(), Error> {
+    if !is_valid(key) || !is_valid_sig(sig) {
+        return Err(Error::InvalidPoint);
+    }
+    let h0m = h0(msg);
+    let p1 = dusk_bls12_381::pairing(sig, &G2Affine::generator());
+    let p2 = dusk_bls12_381::pairing(&h0m, key);
+
+    if p1.eq(&p2) {
+        Ok(())
+    } else {
+        Err(Error::InvalidSignature)
+    }
+}
+
+fn is_valid(key: &G2Affine) -> bool {
+    let is_identity: bool = key.is_identity().into();
+    key.is_torsion_free().into() && key.is_on_curve().into() && !is_identity
+}
+
+/// Aggregated form of a BLS public key.
+/// The public keys are aggregated in a rogue-key attack
+/// resistant manner, by using the hash function defined
+/// in the modified version of BLS.
+#[derive(Default, Copy, Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(
+    feature = "rkyv-impl",
+    derive(Archive, Deserialize, Serialize),
+    archive_attr(derive(bytecheck::CheckBytes))
+)]
+pub struct MultisigPublicKey(G2Affine);
+
+impl Serializable<96> for MultisigPublicKey {
+    type Error = DuskBytesError;
+
+    fn to_bytes(&self) -> [u8; Self::SIZE] {
+        self.0.to_bytes()
+    }
+
+    fn from_bytes(bytes: &[u8; Self::SIZE]) -> Result<Self, Self::Error> {
+        Ok(MultisigPublicKey(G2Affine::from_bytes(bytes)?))
+    }
+}
+
+impl MultisigPublicKey {
+    /// Aggregate a set of [`PublicKey`] into a [`MultisigPublicKey`].
+    ///
+    /// # Errors
+    ///
+    /// The aggregation errors when one of the [`PublicKey`]s is made of the
+    /// identity or an otherwise invalid point.
+    pub fn aggregate(pks: &[PublicKey]) -> Result<Self, Error> {
```

Comment:
Return Error if pks is empty

Reply:
Had to add a new error variant for this, but it makes sense.